src/acl.c

```diff
@@ -937,6 +963,62 @@ static void ACLAddAllowedFirstArg(aclSelector *selector, unsigned long id, const
     selector->allowed_firstargs[id][items - 1] = NULL;
 }
 
+/* Clean up and return error from database permission operations. */
+static int ACLDatabasePermissionError(intset *new_dbs, sds dblist, int error_code) {
+    if (new_dbs) intsetFree(new_dbs);
+    if (dblist) sdsfree(dblist);
+    errno = error_code;
+    return C_ERR;
+}
+
+/* Helper function to add database permissions to a selector. */
+static int ACLSetSelectorDatabasePermissions(aclSelector *selector, const char *dbs_str) {
+    /* Initialize the database intset */
+    intset *new_dbs = intsetNew();
+
+    if (selector->flags & SELECTOR_FLAG_ALLDBS) {
+        selector->flags &= ~SELECTOR_FLAG_ALLDBS;
+    }
+
+    sds dblist = sdsnew(dbs_str);
+    size_t len = sdslen(dblist);
+
+    /* Reject empty list, trailing commas or more than 1 consecutive commas */
+    if (len == 0 || dblist[0] == ',' ||
+        dblist[len - 1] == ',' || strstr(dblist, ",,")) {
+        return ACLDatabasePermissionError(new_dbs, dblist, EINVAL);
+    }
+
+    int count;
+    sds *tokens = sdssplitlen(dblist, len, ",", 1, &count);
+
+    for (int i = 0; i < count; i++) {
+        char *endptr = NULL;
+        errno = 0;
+        int64_t dbid = strtoll(tokens[i], &endptr, 10);
+
+        if (errno == ERANGE || endptr == tokens[i] || *endptr != '\0' ||
+            dbid < 0 || dbid >= server.dbnum) {
```

Comment:
Wouldn't `endptr` not being NULL is sufficient to error out ?

```suggestion
        if (errno == ERANGE || *endptr != '\0' ||
            dbid < 0 || dbid >= server.dbnum) {
```

Reply:
I think we should split into two cases 

1. ERANGE (out of valid database range)
2. EINVAL (encompass invalid input)

WDYT?

Reply:
Good points, did that